lib/backend/test/atomicwrite.go

```diff
@@ -0,0 +1,688 @@
+	defer cancel()
+
+	keyOf := func(i int) []byte {
+		return []byte(fmt.Sprintf("/mkey-%d", i))
+	}
+
+	var condacts []backend.ConditionalAction
+
+	// set up one more conditional actions than should be permitted
+	for i := 0; i < backend.MaxAtomicWriteSize+1; i++ {
+		// delete any dangling state from previous runs of this test
+		if err := bk.Delete(ctx, keyOf(i)); !trace.IsNotFound(err) {
+			require.NoError(t, err)
+		}
+
+		condacts = append(condacts, backend.ConditionalAction{
+			Key:       keyOf(i),
+			Condition: backend.NotExists(),
+			Action: backend.Put(backend.Item{
+				Value: []byte("v1"),
+			}),
+		})
+	}
+
+	// atomic write should fail
+	_, err = bk.AtomicWrite(ctx, condacts...)
+	require.Error(t, err)
+
+	// truncate to the allowed maximum
+	condacts = condacts[:backend.MaxAtomicWriteSize]
+
+	// atomic write should now succeed
+	rev1, err := bk.AtomicWrite(ctx, condacts...)
+	require.NoError(t, err)
+
+	// verify that items were inserted as expected
+	for i := 0; i < backend.MaxAtomicWriteSize; i++ {
+		item, err := bk.Get(ctx, keyOf(i))
+		require.NoError(t, err, "i=%d", i)
+		require.Equal(t, rev1, item.Revision)
+		require.Equal(t, []byte("v1"), item.Value)
+	}
+
+	// update puts
+	for i := range condacts {
+		condacts[i].Action = backend.Put(backend.Item{
+			Value: []byte("v2"),
+		})
+	}
+
+	// re-attempt should fail due to conditions no-longer holding
+	_, err = bk.AtomicWrite(ctx, condacts...)
+	require.ErrorIs(t, err, backend.ErrConditionFailed)
+
+	// verify that failed atomic write results in no changes
+	for i := 0; i < backend.MaxAtomicWriteSize; i++ {
+		item, err := bk.Get(ctx, keyOf(i))
+		require.NoError(t, err, "i=%d", i)
+		require.Equal(t, rev1, item.Revision)
+		require.Equal(t, []byte("v1"), item.Value)
+	}
+
+	// update conditional actions to assert revision
+	for i := range condacts {
+		condacts[i].Action = backend.Put(backend.Item{
+			Value: []byte("v3"),
+		})
+		condacts[i].Condition = backend.Revision(rev1)
+	}
+
+	// conditional actions should now succeed
+	rev2, err := bk.AtomicWrite(ctx, condacts...)
+	require.NoError(t, err)
+
+	// verify that changes occurred as expected
+	for i := 0; i < backend.MaxAtomicWriteSize; i++ {
+		item, err := bk.Get(ctx, keyOf(i))
+		require.NoError(t, err, "i=%d", i)
+		require.Equal(t, rev2, item.Revision)
+		require.Equal(t, []byte("v3"), item.Value)
+	}
+}
+
+// testAtomicWriteConcurrent is a sanity-check intended to verify the correctness of AtomicWrite under high concurrency.
+func testAtomicWriteConcurrent(t *testing.T, newBackend AtomicWriteConstructor) {
+	const (
+		increments = 200
+		workers    = 20
+	)
+	bk, _, err := newBackend()
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	counterKey := []byte("/ccounter")
+
+	_, err = bk.Put(ctx, backend.Item{
+		Key:   counterKey,
+		Value: []byte("0"),
+	})
+	require.NoError(t, err)
+
+	var wg sync.WaitGroup
```

Comment:
If you switched to `golang.org/x/sync/errgroup.WithContext` you could get early exiting and error returning without having to use panics.